File: lib/kronic.rb
require 'date'

# Main class for library, provides date parsing and formatting methods.
class Kronic
  # Public: Converts a human readable day (today, yesterday) to a Date.
  # If Time.zone is available and set (added by active_support and rails),
  # Time.zone.today will be used as a reference point, otherwise Date.today
  # will be used.
  #
  # string - The String to convert to a Date. #to_s is called on the parameter.
  #          Supported formats are: Today, yesterday, tomorrow, last thursday,
  #          this thursday, 14 Sep, Sep 14, 14 June 2010. Parsing is
  #          case-insensitive.
  #
  # Returns the Date, or nil if the input could not be parsed.
  def self.parse(string)
    string = string.to_s.downcase.strip
    now = today

    parse_nearby_days(string, now) ||
      parse_last_or_this_day(string, now) ||
      parse_exact_date(string, now) ||
      parse_iso_8601_date(string)
  end

  # Public: Converts a date to a human readable string. If Time.zone is
  # available and set (it is added by active_support and rails by default),
  # Time.zone.today will be used as a reference point, otherwise Date.today
  # will be used.
  #
  # date - The Date to be converted
  # opts - The Hash options used to customize formatting
  #        :today - The reference point for calculations (default: Date.today)
  #
  # Returns a relative string ("Today", "This Monday") if available, otherwise
  # the full representation of the date ("19 September 2010").
  def self.format(date, opts = {})
    case (to_date(date) - to_date(opts[:today] || today)).to_i
      when (2..7)   then "This " + date.strftime("%A")
      when 1        then "Tomorrow"
      when 0        then "Today"
      when -1       then "Yesterday"
      when (-7..-2) then "Last " + date.strftime("%A")
      else               date.strftime("%e %B %Y").strip
    end
  end

  class << self
    private

    DELIMITER           = /[,\s]+/
    NUMBER              = /^[0-9]+$/
    NUMBER_WITH_ORDINAL = /^([0-9]+)(st|nd|rd|th)?$/
    ISO_8601_DATE       =
      /^([0-9]{4})-?(1[0-2]|0?[1-9])-?(3[0-1]|[1-2][0-9]|0?[1-9])$/

    MONTH_NAMES =
      Date::MONTHNAMES.zip(Date::ABBR_MONTHNAMES).flatten.compact.map {|x|
        x.downcase
      }

    # Ruby 1.8 does not provide a to_date method on Time. This methods works
    # on all rubies.
    def to_date(time)
      Date.new(time.year, time.month, time.day)
    end

    def today
      if Time.respond_to?(:zone) && Time.zone
        to_date(Time.zone.now)
      else
        Date.today
      end
    end

    # Examples
    #
    #   month_from_name("january") # => 1
    #   month_from_name("jan")     # => 1
    def month_from_name(month)
      month = MONTH_NAMES.index(month)
      month ? month / 2 + 1 : nil
    end

    # Parse "Today", "Tomorrow" and "Yesterday"
    def parse_nearby_days(string, today)
      return today     if string == 'today'
      return today - 1 if string == 'yesterday'
      return today + 1 if string == 'tomorrow'
    end

    # Parse "Last Monday", "This Monday"
    def parse_last_or_this_day(string, today)
      tokens = string.split(DELIMITER)

      if %w(last this).include?(tokens[0])
        days = (1..7).map {|x|
          today + (tokens[0] == 'last' ? -x : x)
        }.inject({}) {|a, x|
          a.update(x.strftime("%A").downcase => x)
        }

        days[tokens[1]]
      end
    end

    # Parse "14 Sep", "14 September", "14 September 2010", "Sept 14 2010"
    def parse_exact_date(string, today)
      tokens = string.split(DELIMITER)

      if tokens.length >= 2
        if    tokens[0] =~ NUMBER_WITH_ORDINAL
          parse_exact_date_parts(tokens[0], tokens[1], tokens[2], today)
        elsif tokens[1] =~ NUMBER_WITH_ORDINAL
          parse_exact_date_parts(tokens[1], tokens[0], tokens[2], today)
        end
      end
    end

    # Parses day, month and year parts
    def parse_exact_date_parts(raw_day, raw_month, raw_year, today)
      day   = raw_day.to_i
      month = month_from_name(raw_month)
      year = if raw_year
        raw_year =~ NUMBER ? raw_year.to_i : nil
      else
        today.year
      end

      return nil unless day && month && year

      begin
        result = Date.new(year, month, day)
        result = result << 12 if result > today && !raw_year
        result
      rescue ArgumentError
        nil
      end
    end

    # Parses "2010-09-04", "2010-9-4"
    #
    # NOTE: this is not strictly the ISO 8601 date format as it allows months
    # and days without the zero prefix. e.g. 2010-9-4
    def parse_iso_8601_date(string)
      if string =~ ISO_8601_DATE
        Date.parse(string) rescue nil
      end
    end
  end
end

File: lib/kronic.js
var Kronic = (function() { 
  function trim(string) {
    return string.replace(/^\s+|\s+$/g, '')
  }

  function map(array, func) {
    var result = [];
    for (x in array) {
      result.push(func(array[x]));
    }
    return result;
  }
  
  function inject(array, initialValue, func) {
    var accumulator = initialValue;
    for (x in array) {
      accumulator = func(accumulator, array[x]);
    }
    return accumulator;
  }

  function addDays(date, numberOfDays) {
    return new Date(date * 1 + numberOfDays * 60 * 60 * 24 * 1000);
  }

  function parseNearbyDays(string, today) {
    if (string == 'today')     return today;
    if (string == 'yesterday') return addDays(today, -1);
    if (string == 'tomorrow')  return addDays(today, +1);
  }

  function parseLastOrThisDay(string, today) {
    var tokens = string.split(/\s+/);

    if (['last', 'this'].indexOf(tokens[0]) >= 0) {
      var days = map([1,2,3,4,5,6,7], function(x) {
        return addDays(today, tokens[0] == 'last' ? -x : x);
      });

      days = inject(days, {}, function(a, x) {
        a[x.strftime("%A").toLowerCase()] = x;
        return a;
      });

      return days[tokens[1]];
    }
  }

  function parseExactDay(string, today) {
    var tokens = string.split(/\s+/);

    if (tokens.length >= 2) {
      var matches = tokens[0].match(/^([0-9]+)(st|nd|rd|th)?$/);
      if (matches) {
        return parseExactDateParts(matches[1], tokens[1], tokens[2], today);
      } else {
        matches = tokens[1].match(/^([0-9]+)(st|nd|rd|th)?$/);
        return parseExactDateParts(matches[1], tokens[0], tokens[2], today);
      }
    }
  }

  function parseExactDateParts(rawDay, rawMonth, rawYear, today) {
    var day = rawDay * 1;
    var month = monthFromName(rawMonth);
    var year;

    if (rawYear)
      year = rawYear.match(/^[0-9]+$/) ? rawYear * 1 : null;
    else
      year = today.getYear() + 1900;

    if (!(day && month && year))
      return null;

    var result = new Date(year, month, day);
    if (result > today && !rawYear)
      result = new Date(year - 1, month, day);
    return result;
  }

  function parseIso8601Date(string) {
    if (string.match(/^([0-9]{4})-?(1[0-2]|0?[1-9])-?(3[0-1]|[1-2][0-9]|0?[1-9])$/)) {
      var tokens = map(string.split('-'), function(x) { return x * 1});
      return (new Date(tokens[0], tokens[1] - 1, tokens[2]));
    }
  }

  function monthFromName(month) {
    var monthNames = [];
    for (var i = 0; i < 12; i++) {
      monthNames.push((new Date(2010, i, 1)).strftime("%b").toLowerCase());
      monthNames.push((new Date(2010, i, 1)).strftime("%B").toLowerCase());
    }
    monthIndex = monthNames.indexOf(month);
    return monthIndex >= 0 ? Math.floor(monthIndex / 2) : null;
  }

  return {
    parse: function(string) {
      string = trim(string + '').toLowerCase();
      return parseNearbyDays(string, Kronic.today()) ||
             parseLastOrThisDay(string, Kronic.today()) ||
             parseExactDay(string, Kronic.today()) ||
             parseIso8601Date(string);
    },
    format: function(date, opts) {
      if (!opts)
        opts = {today: Kronic.today()};

      var diff = Math.floor((date * 1 - opts.today * 1) / 60 / 60 / 24 / 1000);
      switch (diff) {
        case -7:
        case -6:
        case -5:
        case -4:
        case -3:
        case -2: return "Last " + date.strftime("%A");
        case -1: return "Yesterday";
        case 0:  return "Today";
        case 1: return "Tomorrow";
        case 2:
        case 3:
        case 4:
        case 5:
        case 6:
        case 7: return "This " + date.strftime("%A");
        default: return date.strftime("%e %B %Y");
      }
    },
    today: function() {
      return new Date();
    }
  };
})();
